### ISSUE TITLE
Testing: spatial branch (new geometries, et al.)

### ISSUE THREAD

**Issue by jimsafley** [omeka-s-modules/Mapping#117]
The `spatial` branch addresses a handful of issues, including the addition of some significant features:

- Adds support for polygon, rectangle, and polyline geometries ([#93](https://github.com/omeka-s-modules/Mapping/issues/93))
- Adds a site setting to disable feature clustering ([#96](https://github.com/omeka-s-modules/Mapping/issues/96))
- Adds a site setting for site-wide basemap provider ([#112](https://github.com/omeka-s-modules/Mapping/issues/112))
- Adds a map to the admin item set show page and the public item set show page (via resource page block) ([#106](https://github.com/omeka-s-modules/Mapping/issues/106))
- Enables easier modification of feature pop-up markup (via template overwrite) ([#68](https://github.com/omeka-s-modules/Mapping/issues/68))
- Standardizes feature pop-up markup ([#113](https://github.com/omeka-s-modules/Mapping/issues/113))
- Popup and timeline titles respect site language ([#115](https://github.com/omeka-s-modules/Mapping/issues/115))

**Important**: before testing you should have the most recent, published version of Mapping already installed and a sizable collection of markers assigned to multiple items. This is necessary to test the migration from "markers" to "features." Once you set this up, you can continue with the test.

To test the module, you'll need to pull in the latest changes in the Omeka S `develop` branch, and checkout Mapping's `spatial` branch. You must install dependencies by running the following command in the /modules/Mapping/ directory:

```
$ gulp deps:module
```

After this successfully runs, go to the modules page and upgrade Mapping. If this works, you can continue testing.

Given the extensive changes to the code, testing the branch must involve a wholesale evaluation of every component in the module. Save for the new features, everything should work as before:

- Item map form (make sure the UI works)
- Item show, admin and public (make sure bounds and geographical features are accurate)
- Item set show, admin and public (make sure geographical features are accurate)
- Advanced search (test "feature presence" and "geographic location")
- Batch update (test "copy coordinates" and "update features")
- Site blocks (test "map by attachment" and "map by query", including WMS, timeline, etc.)
- Site navigation link (map browse) (make sure geographical features are accurate and search works)
- Site settings (test "add feature presence," "add geographic location," and "disable feature clustering")
- Module integrations:
    - CSV Import (CSVImport)
    - Omeka Classic Importer (Omeka2Importer)
    - Omeka S Item Importer (Osii)
    - Collecting

If for any reason you need to downgrade to the previous version, you must uninstall Mapping, checkout Mapping's `master` branch, and reinstall the module. In fact, it's probably worth the effort to test the migration several times.

Comment here if you find a bug or unexpected behavior, or if anything needs more clarity. If you have additional feature requests, please compile them until after testing is over and then we can discuss them on a separate issue.

**Comment by allanaaa** [comment]
I'm getting started on this. 
What's the intended relationship between the automated clusters of markers and a shape? 

<img width="1155" alt="Screenshot 2023-09-18 at 1 14 39 PM" src="https://github.com/omeka-s-modules/Mapping/assets/28012148/da743cfd-3081-4d25-ac11-056eb594ba89">

Right now I'm hovering over the "76" to make its cluster polygon visible, so you can see it overlaps with a manual rectangle above and to the left of it. 

I think the first thing to do is pick a second colour combo to differentiate shapes/lines and clusters. Probably don't need to put them in a little legend, but I'll leave that up to you. 

That might be enough, but I think if people have clusters enabled they might want the shapes/lines to cluster as well at certain levels. Then we can debate whether it should be when they overlap a little bit, or entirely ....

<img width="1018" alt="Screenshot 2023-09-18 at 1 17 47 PM" src="https://github.com/omeka-s-modules/Mapping/assets/28012148/d2f6be22-40c2-453d-a4f5-ae691f4f420c">

The same map fully zoomed out, hovering over the "255" cluster. I think you can see that this will be a problem if, say, you have an identical polgyon over a whole country for multiple items. They will need to cluster. 

So, speaking of "identical shapes", what will be the CSV Import integration to input these shapes? I don't see anything new in the dropdown now but I imagine you will allow a series of GPS points to make lines and polygons, and a right/top/left/bottom for the rectangle, same as the way default bounds works now. Will there be a maximum number of points before we start seeing a lag? Or a maximum number of map objects per item?

**Comment by jimsafley** [comment]
It's a good idea to differentiate between cluster polygons and user polygons. I've changed cluster polygons to green.

Are your shapes not clustering? The logic is opaque, but above a certain threshold (defined by zoom level, shape size, and shape position, I'm assuming) shapes should cluster along with markers. It could be that your shapes don't meet that threshold?

As for CSV Import: I'm not going to address CSV integration until after testing and likely after release. I want to make sure the i/o format is stable before doing integrations.

**Comment by allanaaa** [comment]
Yep, you're right, smaller shapes are clustering nicely! It's only when we get to country-sized ones (or province, or Alaska) that they never cluster. 

We'll end up with something like:
<img width="971" alt="Screenshot 2023-09-18 at 8 31 32 PM" src="https://github.com/omeka-s-modules/Mapping/assets/28012148/cd16eff1-e662-42e0-b448-e1587fe28fda">

Where one item pretty much totally covers the other one, even though you can sort of see there are two shapes there. Can you get the clustering to happen at a different threshold?

**Comment by jimsafley** [comment]
> Can you get the clustering to happen at a different threshold?

No. I can modify the threshold for markers but not for other features. The JS library responsible for clustering does not natively support shapes. We use another library to integrate shapes into clusters, but it's not optimized to the degree needed to configure threshold.

Beyond clustering, have you encountered other issues? I'm particularly interested in the migration. Of all the things that must be working before release, migration is the most important.

**Comment by allanaaa** [comment]
> * Adds support for polygon, rectangle, and polyline geometries (#93)

Seems good, they draw easily and save and display well. 

> * Adds a site setting for site-wide basemap provider (#112)

Looks good, you can set a site-wide and override it in individual page blocks and on the map browse page; turning off the override and back to the default worked.

> * Adds a site setting to disable feature clustering (#96)

Works fine, but doesn't it make more sense to enable with a check, rather than disable with a check? 

I don't like the language of "feature". It's ambiguous and already used in a number of other contexts. If you want a term to encompass markers + shapes + lines, we should use "map objects" or "coordinates" or "geolocators" or "locations". 

May also need some interface text to explain some shapes won't cluster because they are too big. 

> * Adds a map to the admin item set show page and the public item set show page (via resource page block) (#106)

Good, though I worry people will want item-set-specific map customization settings next. 



I am getting an issue where I have multiple map blocks on the homepage, one map-by-query showing every item, and another map-by-attachment with only one item attached. But they're both showing all items. 
I've just added a third with only one item attached, including a timeline, and it's also showing all items. All three have separate basemaps selected so some code is generating uniquely, but not the items for each. Rearranging them doesn't seem to govern which setting gets used on all maps.

**Comment by jimsafley** [comment]
> Works fine, but doesn't it make more sense to enable with a check, rather than disable with a check?

Typically a default setting is unchecked. Clustering is on by default, so turning it off is opt-in, therefore checked.

> I don't like the language of "feature". It's ambiguous and already used in a number of other contexts. 

"Feature" is so commonly used in GIS that I'm reluctant to use anything else. Even so, I'm going to change the label to "Disable clustering of map features" and add an unambiguous description of what feature means in this context.

> May also need some interface text to explain some shapes won't cluster because they are too big.

Will do.

> I am getting an issue where I have multiple map blocks on the homepage, one map-by-query showing every item, and another map-by-attachment with only one item attached. But they're both showing all items.

Yes, that's a bug. Easy fix.

**Comment by sharonmleon** [comment]
Perhaps an update error, but my map browse page is posting an error: 

http://6floors.org/workspace/omeka-s/s/photographs/map-browse

**Comment by zerocrates** [comment]
There's a call to an overridden search form partial from Freedom in that stack. Does the same error happen if you're not using Freedom?

**Comment by sharonmleon** [comment]
Yes it happens with updated versions of all the themes.

**Comment by jimsafley** [comment]
Could you be using a modified version of /view/mapping/site/index/browse.phtml? The stack says the partial call is on line 28, but the call is actually on line 22.

**Comment by sharonmleon** [comment]
How would I know that? Would I have edited that myself? I haven't done any code insertions to my knowledge.

**Comment by allanaaa** [comment]
I'm getting this error on Freedom but not the other themes in the same install:

> Notice: Undefined variable: markers in /var/www/html/amayer/amayer-s/omeka-s/themes/freedom/view/mapping/site/index/browse.phtml on line 31

> Warning: Invalid argument supplied for foreach() in /var/www/html/amayer/amayer-s/omeka-s/themes/freedom/view/mapping/site/index/browse.phtml on line 31

http://dev.omeka.org/amayer/amayer-s/omeka-s/s/my-third-site/map-browse

**Comment by jimsafley** [comment]
The Freedom theme overrides the map browse page, which has yet to be updated to work with Mapping's spatial branch. Those warnings will disappear once its updated.

**Comment by jimsafley** [comment]
Note that I've merged the `spatial` branch into `master`, so you'll need to do further testing on that.
